File: controllers/ssoSettingsController.js
const SSOSettings = require('../models/SSOSettings')

exports.getAllSSoSettings = async (req, res) => {
    try {
        const settings = await SSOSettings.find();
        res.json(settings);
    } catch (err) {
        res.status(500).json({ message: err.message });
    }
}

// Get a specific SSO setting by provider
exports.getAllSSOSettingsByProvider = async (req, res) => {
    try {
        const settings = await SSOSettings.findOne({ provider: req.params.provider });
        if (!settings) {
            return res.status(404).json({ message: 'Provider settings not found' });
        }
        res.json(settings);
    } catch (err) {
        res.status(500).json({ message: err.message });
    }
}


// Create or update SSO settings
exports.saveSSOSettings = async (req, res) => {
    const { provider = 'google', loginUrl, redirectUrl, clientId, tenantId, secret, secretExpiry, scopes } = req.body;

    try {
        // Check if settings for the provider already exist
        const existingSettings = await SSOSettings.findOne({ provider });

        if (existingSettings) {
            // Update existing settings
            existingSettings.loginUrl = loginUrl || existingSettings.loginUrl;
            existingSettings.redirectUrl = redirectUrl;
            existingSettings.clientId = clientId;
            existingSettings.tenantId = tenantId;
            existingSettings.clientSecret = secret;
            existingSettings.secretExpiry = secretExpiry;
            existingSettings.scopes = scopes;

            const updatedSettings = await existingSettings.save();
            return res.status(200).json({ message: 'Settings updated', data: updatedSettings });
        }

        // Create new settings
        const newSettings = new SSOSettings({
            provider,
            loginUrl,
            redirectUrl,
            clientId,
            tenantId,
            clientSecret: secret,
            secretExpiry,
            scopes,
        });

        const savedSettings = await newSettings.save();
        res.status(201).json({ message: 'Settings saved', data: savedSettings });
    } catch (err) {
        res.status(500).json({ message: err.message });
    }
};

// Delete SSO settings
exports.deletedSettings = async (req, res) => {
    try {
        const deletedSettings = await SSOSettings.findOneAndDelete({ provider: req.params.provider });
        if (!deletedSettings) {
            return res.status(404).json({ message: 'Provider settings not found' });
        }
        res.json({ message: 'Settings deleted', data: deletedSettings });
    } catch (err) {
        res.status(500).json({ message: err.message });
    }
};

File: controllers/sso_settings_controller.py
from fastapi import HTTPException
from models import SSOSettings  # Assuming Beanie model
from typing import Dict, Any, List

class SSOSettingsController:
    @staticmethod
    async def get_all_sso_settings() -> List[Dict[str, Any]]:
        """Retrieve all SSO settings."""
        try:
            settings = await SSOSettings.find_all().to_list()
            return [setting.dict() for setting in settings]
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @staticmethod
    async def get_all_sso_settings_by_provider(provider: str) -> Dict[str, Any]:
        """Retrieve SSO settings by provider."""
        try:
            settings = await SSOSettings.find_one(SSOSettings.provider == provider)
            if not settings:
                raise HTTPException(status_code=404, detail="Provider settings not found")
            return settings.dict()
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @staticmethod
    async def save_sso_settings(data: Dict[str, Any]) -> Dict[str, Any]:
        """Create or update SSO settings."""
        provider = data.get("provider", "google")
        try:
            existing_settings = await SSOSettings.find_one(SSOSettings.provider == provider)
            if existing_settings:
                # Update existing settings
                for key, value in data.items():
                    if value is not None:
                        setattr(existing_settings, key, value)
                await existing_settings.save()
                return {"message": "Settings updated", "data": existing_settings.dict()}
            else:
                # Create new settings
                new_settings = SSOSettings(**data)
                await new_settings.insert()
                return {"message": "Settings saved", "data": new_settings.dict()}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @staticmethod
    async def delete_settings(provider: str) -> Dict[str, Any]:
        """Delete SSO settings by provider."""
        try:
            settings = await SSOSettings.find_one(SSOSettings.provider == provider)
            if not settings:
                raise HTTPException(status_code=404, detail="Provider settings not found")
            await settings.delete()
            return {"message": "Settings deleted", "data": settings.dict()}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))